### PR TITLE
Add horizon-cmp command

### PR DESCRIPTION
**Pull request by bartekn**
Adds a simple command to compare responses of two horizon servers. Can be used to check differences between the latest stable version and release candidate.

EDIT: forgot about "Work in Progress" option in GH, can't change the status once PR is added.

TODO:
- [x] Add more paths to the test corpus (can also add code that generates relevant paths).
- [x] Print diff instead of full body when responses differ.
- [x] ~Setting params via CLI.~ #1183
- [x] ~Test streaming.~ #1184
- [x] ~Run in CI and add diffs to PRs.~ #1185

### CODE REVIEW

File: tools/horizon-cmp/internal/response.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+		panic(err)
+	}
+
+	response.Body = string(body)
+
+	normalizedBody := response.Body
+	// `result_meta_xdr` can differ between core instances (confirmed this with core team)
+	normalizedBody = findResultMetaXDR.ReplaceAllString(normalizedBody, "")
+	// Remove Horizon URL from the _links
+	normalizedBody = strings.Replace(normalizedBody, domain, "", -1)
```
Comment (howardtw) [inline tools/horizon-cmp/internal/response.go]: Maybe we can use `strings.ReplaceAll`?

File: tools/horizon-cmp/internal/response.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		panic(err)
+	}
+
+	response.Body = string(body)
+
+	normalizedBody := response.Body
+	// `result_meta_xdr` can differ between core instances (confirmed this with core team)
+	normalizedBody = findResultMetaXDR.ReplaceAllString(normalizedBody, "")
```
Comment (howardtw) [inline tools/horizon-cmp/internal/response.go]: maybe we can inline this?
```
normalizedBody = findResultMetaXDR.ReplaceAllString(response.Body, "")
```

File: tools/horizon-cmp/internal/response.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+func (r *Response) SaveDiff(outputDir string, other *Response) {
+	if r.Path != other.Path {
+		panic("Paths are different")
+	}
+
+	fileName := pathToFileName(r.Path)
+
+	if len(fileName) > 100 {
```
Comment (howardtw) [inline tools/horizon-cmp/internal/response.go]: why is it a hundred?

File: tools/horizon-cmp/internal/response.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	err = ioutil.WriteFile(fileB, []byte(other.Path+"\n\n"+other.Body), 0744)
+	if err != nil {
+		panic(err)
+	}
+
+	out, err := exec.Command("diff", fileA, fileB).Output()
+	if err != nil {
+		// Ignore, user will generate diff manually
```
Comment (howardtw) [inline tools/horizon-cmp/internal/response.go]: Why do we want to ignore the error returned by the `Output` given that the user will generate `diff` program manually? And if it makes more sense to ignore the error, maybe we can just do:
```
out, _ := exec.Command("diff", fileA, fileB).Output()
```

File: tools/horizon-cmp/internal/response.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	for _, link := range found {
+		l := strings.Replace(string(link[1]), "\\u0026", "&", -1)
+		links = append(links, l)
+	}
+
+	return links
+}
+
+func pathToFileName(path string) string {
+	path = strings.Replace(path, "/", "_", -1)
```
Comment (howardtw) [inline tools/horizon-cmp/internal/response.go]: Can we use `strings.ReplaceAll` here or they are different?

File: tools/horizon-cmp/internal/response.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	path = strings.Replace(path, "?", "_", -1)
+	path = strings.Replace(path, "&", "_", -1)
+	path = strings.Replace(path, "=", "_", -1)
```
Comment (howardtw) [inline tools/horizon-cmp/internal/response.go]: we can just return here :)

File: tools/horizon-cmp/main.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+)
+
+const HorizonOld = "https://horizon.stellar.org"
+const HorizonNew = "http://localhost:8001"
+
+const MaxLevels = 3
+
+type PathWithLevel struct {
+	Path  string
+	Level int
```
Comment (howardtw) [inline tools/horizon-cmp/main.go]: what does `Level` mean here?

File: tools/horizon-cmp/main.go [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	protocol "github.com/stellar/go/protocols/horizon"
+	cmp "github.com/stellar/go/tools/horizon-cmp/internal"
+)
+
+const HorizonOld = "https://horizon.stellar.org"
```
Comment (howardtw) [inline tools/horizon-cmp/main.go]: maybe we can unexport these constants?

File: tools/horizon-cmp/internal/response.go (hunk 1 above)
Comment (bartekn) [reply tools/horizon-cmp/internal/response.go]: It was added in Go 1.12. We compile using Go 1.11.

File: tools/horizon-cmp/internal/response.go (hunk 3 above)
Comment (bartekn) [reply tools/horizon-cmp/internal/response.go]: Moved to a `const`. Basically file system limits.

File: tools/horizon-cmp/internal/response.go (hunk 4 above)
Comment (bartekn) [reply tools/horizon-cmp/internal/response.go]: Done.

File: tools/horizon-cmp/internal/response.go (hunk 5 above)
Comment (bartekn) [reply tools/horizon-cmp/internal/response.go]: https://github.com/stellar/go/pull/977#discussion_r278553399

File: tools/horizon-cmp/main.go (hunk 7 above)
Comment (bartekn) [reply tools/horizon-cmp/main.go]: Added explanation to `const maxLevel`.

File: tools/horizon-cmp/main.go (hunk 8 above)
Comment (bartekn) [reply tools/horizon-cmp/main.go]: Good point, unexported all variables.

**Comment by bartekn** [comment]
@howardtw can you take a look again?

File: tools/horizon-cmp/internal/response.go (hunk 2 above)
Comment (howardtw) [reply tools/horizon-cmp/internal/response.go]: I think this is left out?

File: tools/horizon-cmp/internal/response.go (hunk 6 above)
Comment (howardtw) [reply tools/horizon-cmp/internal/response.go]: And this one too

File: tools/horizon-cmp/main.go [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+func init() {
+	visitedPaths = make(map[string]bool)
+}
+
+func main() {
+	// Get latest ledger and operate on it's cursor to get responses at a given ledger.
+	ledger := getLatestLedger()
+	cursor := ledger.PagingToken()
+
+	// Sleep for a few sec to make sure second Horizon is up to speed
```
Comment (howardtw) [inline tools/horizon-cmp/main.go]: you mean `Sleep for a few seconds to make sure the second Horizon is up to speed?`

File: tools/horizon-verify/main.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+	"github.com/stellar/go/xdr"
+)
+
+var horizonURL string
+var startSequence uint32
+var count uint
+
+func init() {
+	rootCmd.PersistentFlags().StringVarP(&horizonURL, "url", "u", "", "Horizon server URL")
+	rootCmd.PersistentFlags().Uint32VarP(&startSequence, "start", "s", 0, "Sequence number of a start ledger (follows descending order, defaults to the latest ledger)")
```
Comment (howardtw) [inline tools/horizon-verify/main.go]: I'm not sure if this reads better `Sequence number of the ledger to start with`

File: tools/horizon-verify/main.go [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+					ForLedger:     uint(ledger.Sequence),
+					Limit:         200,
+					IncludeFailed: true,
+				})
+
+				if err != nil {
+					panic(err)
+				}
+
+				var wg sync.WaitGroup
```
Comment (howardtw) [inline tools/horizon-verify/main.go]: nit: maybe we can put these variables together
```
var(
   wg sync.WaitGroup
   successful, failed int32
)
```

File: tools/horizon-verify/main.go [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+						failed++
+					}
+
+					go func(transaction protocol.Transaction) {
+						defer wg.Done()
+
+						var resultXDR xdr.TransactionResult
+						err = xdr.SafeUnmarshalBase64(transaction.ResultXdr, &resultXDR)
+						if err != nil {
+							return
```
Comment (howardtw) [inline tools/horizon-verify/main.go]: should we log this err or panic?

File: tools/horizon-verify/main.go [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+						}
+
+						if transaction.Successful && resultXDR.Result.Code != xdr.TransactionResultCodeTxSuccess {
```
Comment (howardtw) [inline tools/horizon-verify/main.go]: maybe we can combine the two if cases `if (transaction.Successful && resultXDR.Result.Code != xdr.TransactionResultCodeTxSuccess) || (!transaction.Successful && resultXDR.Result.Code == xdr.TransactionResultCodeTxSuccess)`

File: tools/horizon-cmp/main.go (hunk 9 above)
Comment (bartekn) [reply tools/horizon-cmp/main.go]: Done.

File: tools/horizon-verify/main.go (hunk 10 above)
Comment (bartekn) [reply tools/horizon-verify/main.go]: Done.

File: tools/horizon-verify/main.go (hunk 11 above)
Comment (bartekn) [reply tools/horizon-verify/main.go]: Done.

File: tools/horizon-verify/main.go (hunk 12 above)
Comment (bartekn) [reply tools/horizon-verify/main.go]: Good point, changed.

File: tools/horizon-verify/main.go (hunk 13 above)
Comment (bartekn) [reply tools/horizon-verify/main.go]: Done.

**Comment by bartekn** [comment]
Not sure why I can't reply to the first two comments (no text box). Here they are:

> https://github.com/stellar/go/pull/977#discussion_r278693673

We need to remove `result_meta_xdr`because it can be different between stellar-core instances, unfortunately. It's using `unordered_map` so the order can be different.

> https://github.com/stellar/go/pull/977#discussion_r278694593

This is the code that generates file name from the path. We can't remove it.

**Pull request merged**